README
Edit the file url.js and change the settings at the top, being careful not to alter the code itself.

There are several other things you can do to improve how this works with your OpenSim instance.

(1) On a Linux/Unix server, make a soft link from the [opensim]/bin/maptiles/[UUID] folder ("00000000-0000-0000-0000-000000000000" by 
default) to the [google_map]/data/region folder, substituting the path name for [opensim] and [google_map], wherever these are on your 
server and whatever you are using for [UUID]. This will prevent you from having to manually import the maptiles into the folder, with 
the added bonus that they will be updated automatically.

You could link the files inside the region folder individually but this is a hassle because you need to do it each time you add regions. If 
the regions are on another server then you are going to need to get copies of them and put them in an appropriate place or link them manually 
in some appropriate manner. You will need one for each of the regions, and will need to add others each time you add regions. You should
ignore any files starting in map-2- or more, as only the map-1- files are on the correct zoom level. Filenames in [opensim]/bin/maptiles
that are in the form [UUID].jpg are not local regions on your server, so you should not use these (you must have visited these by
hypergrid, if they exist). You should use the syntax:

ln -s [opensim]/bin/maptiles/map-1-[x]-[y]-objects.jpg [google_map]/data/regions/maptiles/map-1-[x]-[y]-objects.jpg

Where [x] and [y] are the grid locations (not multiplied by 256 as elsewhere).

e.g. ln -s /opt/opensim/bin/maptiles/map-1-1000-1000-objects.jpg /var/www/opensim/app/google_map/data/regions/maptiles/map-1-1000-1000-objects.jpg

Another possible method would be to script this automatically if you parsed the XML output of [google_map]/data/map.php. If you have a great 
deal of regions, especially if people other than you add them or remove them frequently, this might be desirable - it's otherwise unnecessary.

Note that this will only work if you use the option var <code> filenames = "from-files"; </code> instead of <code> var filenames = "uuid";
</code> or <code> var filenames = "uuid-no-dashes"; </code> in the settings. (Note that the tags <code> and </code> are just used here
for clarity and don't appear in the file!) There is a new setting  <code> filenames = "from-server"; </code> but be aware that OpenSim does 
not yet use https (though this may change from dev version 0.9.1.0) and you may get mixed content problems. It obviates the need for the 
symbolic links but introduces some new problems with secure sites.

(2) Use the module Warp3DImageModule rather than MapImageModule. You actually need these settings under [Startup] in OpenSim.ini (as of 
OpenSim 0.7.3.1) - the quotation marks may or may not be critical. Some may already be set in OpenSimDefaults.ini, which you should NOT 
change. If you wish to change settings, copy the lines to the correct sections in OpenSim.ini and change them there:

MapImageModule = "Warp3DImageModule"
TextureOnMapTile = true
DrawPrimOnMapTile = true

(See http://opensimulator.org/wiki/Warp3DImageModule for further details.)

This will dramatically improve the quality of your map tiles but some people have reported that it places excessive demands on memory. It 
is likely that this will depend on the specifications of your server. Check the OpenSim Mantis (http://opensimulator.org/mantis/main_page.php) 
if you have a problem, as someone else may have had it before and you shouldn't duplicate error reports. Likely they are working on it :-)

(3) Notice that, unlike the previous version, you can add multiple map centres (see the settings in url.js), which can even be on different 
servers, providing of course that you have local copies of maptiles or can fetch them somehow from any remote servers when you need to (for 
example, it would be possible to do this automatically using scp or similar if you needed to script this for a large grid, updating at 
intervals using a cron job, for example, on a Linux/Unix/Solaris server). You should be able to have quite a lot, if you *really* need to! ;-)

This opens the way for people to run grids where not all of the regions are in the same general area of the grid, as you can now see them on
your world map (although you could always do this if you didn't worry about seeing them on your web site). This is useful, for example,
if you want to make hypergrid jumps avoiding the 4096 region maximum distance bug in the viewer code. Rather than jumping via an external
intermediary grid in a known middle location, it's much more reliable and predictable to be able to do it in your own home grid. You could
maybe have one set of regions near (7000,7000) and more somewhere like (4000,4000) and you'd be able to reach most other grids. It was formerly 
a bad idea to choose x and y coordinates separated by more than 4096, e.g. (10000,3000), but this may only be an issue for older viewers.

The new code in 2014 makes a number of minor improvements as well as the major new rewrite to support larger regions e.g. varregions etc.

Perhaps you have more than one grid, or you have regions on more than one large public grid like OSGrid, FrancoGrid etc? Now you can put them 
on your own map, at different map centres, providing that you have the map tiles (256 x 256px) available :-)

CHANGELOG
Google maps API v3 interface for OpenSim. (See demo at http://talatchaudhri.com/demo/opensimmaps/v3/)
Based on pre-existing (2008) Google maps API v2 interface for OpenSim. (See demo at http://talatchaudhri.com/demo/opensimmaps/v2/)

Given that GMaps v2 is deprecated and will be withdrawn in 2013, upgraded to v3. As the v3 API is radically different, only a core of the code was portable, so wrote substantial new replacement code.

Various bug fixes and maintenance e.g. map centre, local defaults, links and link protocols.
Re-wrote the info window script and added a draggable/clickable marker, limited within map tile boundaries.
Added multiple centre locations and/or servers.
Added multiple map data copyright support for multiple servers.
Added home button functionality missing in GMaps v3 controls.
Added more configurable user settings than previously added to v2 version.
Added: user setting for initial zoom size.
Support for uuid-based and opensim-based tile filenames (can update automatically from OpenSim server).
Created fake PHP5 DOM script returning XML to enable test instance independent of OpenSim server live XML.
Back-ported DOM script to PHP4 (using older php_domxml) so it can run on PHP4.
Fixed: minor marker setup bug.
Cleaned up code and added comments for documentation.
Added README.
Unfixable (minor): marker hover incorrect on drag if not moved off marker and returned: no JS solution, apparently a browser issue.
Re-wrote to support large regions e.g. varregions in OpenSim/Aurora/WhiteCore
Added offsets from map centre
Improved home centering per map centre
Added zoom levels for each  map centre
Added automatic copyright start/end years
Provided easier mechanism of getting maptiles automatically from the server rather than requiring symbolic links, with setting option to revert to the old method.


To do (features):
add switchable grid map background with grid locations (& sort out how background is done)
move set up user settings into XML and use parser to initialise

Code base last modified: 2019-10-07
This change log last modified: 2019-10-07

index.php
<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.0 Strict//EN" "http://www.w3.org/TR/xhtml1/DTD/xhtml1-strict.dtd">
<html xmlns="http://www.w3.org/1999/xhtml">
 	<head>
   		<meta http-equiv="content-type" content="text/html; charset=utf-8"/>
    	<title>World Map</title>
		<script type="text/javascript" src="http://maps.google.com/maps/api/js?sensor=false&v=3&key=AIzaSyBACCLjQjfliUdoyI90ZS5HNf7M22TYORI"></script>
    	<script type="text/javascript" src="url.js"></script>
        <script type="text/javascript" src="copyright.js"></script>
    	<link rel="stylesheet" type="text/css" href="style.css" />

	</head>
 	<body onload="load()" onunload="GUnload()">
		<div id="content">
			<table><tr><td align="center"><div id="map-canvas" style="width:100%;height:100%;"></div></td></tr></table>
		</div>
  </body>
</html>


data/map.php
<?
include("../../../settings/config.php");
include("../../../settings/mysql.php");

//Creates XML string and XML document using the DOM 
$dom = new DomDocument('1.0', "UTF-8"); 

$map = $dom->appendChild($dom->createElement('Map')); 

$DbLink = new DB;
$DbLink->query("SELECT uuid,regionName,locX,locY,sizeX,sizeY FROM ".C_REGIONS_TBL);
	while(list($UUID,$regionName,$locX,$locY,$dbsizeX,$dbsizeY) = $DbLink->next_record())
	{
	$grid = $map->appendChild($dom->createElement('Grid')); 

	$uuid = $grid->appendChild($dom->createElement('Uuid')); 
	$uuid->appendChild($dom->createTextNode($UUID)); 

	$region = $grid->appendChild($dom->createElement('RegionName')); 
	$region->appendChild($dom->createTextNode($regionName)); 

	$locationX = $grid->appendChild($dom->createElement('LocX')); 
	$locationX->appendChild($dom->createTextNode($locX/256)); 

	$locationY = $grid->appendChild($dom->createElement('LocY')); 
	$locationY->appendChild($dom->createTextNode($locY/256)); 

        $sizeX = $grid->appendChild($dom->createElement('SizeX')); 
        $sizeX->appendChild($dom->createTextNode($dbsizeX)); 

        $sizeY = $grid->appendChild($dom->createElement('SizeY')); 
        $sizeY->appendChild($dom->createTextNode($dbsizeY)); 
	}

$dom->formatOutput = true; // set the formatOutput attribute of 
                            // domDocument to true 
// save XML as string or file 
$test1 = $dom->saveXML(); // put string in test1 
//echo $test1;
header("Content-type: text/xml");
echo $test1;

?>


url2.js
var __items;
var xstart = 10000;
var ystart = 10000;
var hgdomain = 'opensim.example.com';
var hgport = '8002';
var xjump;
var yjump;

function load(){
	if (GBrowserIsCompatible()){
		var div = document.getElementById("map");
		if(div==null){return;}
		if(window.innerWidth){
			div.style.width  = (window.innerWidth - 20) + "px";
			div.style.height = (window.innerHeight - 30) + "px";
		}else{
			div.style.width  = (document.documentElement.clientWidth - 20) + "px";
			div.style.height = (document.documentElement.clientHeight - 30) + "px";			
		}
		var copyCollection = new GCopyrightCollection('<span class="cpy">World Map</span>');
		var copyright = new GCopyright(1, new GLatLngBounds(new GLatLng(0, 0), new GLatLng(20, 10)), 0, "<span class='cpy'>&copy; 2008 OSgrid.org</span>");
		copyCollection.addCopyright(copyright);
		var tilelayers = [new GTileLayer(copyCollection, 5, 9)];
		tilelayers[0].getTileUrl = function CustomGetTileUrl(a,b){
			return "default.jpg";
		}
		var custommap = new GMapType(tilelayers, new GMercatorProjection(10), "World Map", {errorMessage:"no data"});
		var map = new GMap2(div)
		map.addControl(new GLargeMapControl());
		map.enableScrollWheelZoom(); 
		map.addMapType(custommap);
		map.removeMapType(G_NORMAL_MAP); 
		map.removeMapType(G_SATELLITE_MAP); 
		map.removeMapType(G_HYBRID_MAP); 
		map.setCenter(new GLatLng(11,11),8);
		GEvent.addListener(map,"click", function(overlay,point){   
				if(overlay){return;}
				var x = point.lng();
				xjump = Math.round(256 * (x - (x | 0)));
				if(x < 0) x--;		
				var str = x.toString();		
				str = str.substring(0, str.indexOf(".", 0));
				x = xstart - 10 + parseInt(str);
				var y = point.lat();			
				yjump = Math.round(256 * (y - (y | 0)));
				if(y < 0) y--;	
				str = y.toString();
				str = str.substring(0, str.indexOf(".", 0));
				y = ystart - 10 + parseInt(str);	
				if(isOutOfBounds(x,y)){return;}
				//show info popup if a region exist
				var content = getRegionInfos(x,y);
				if(content!=""){
					msg = content;	
					map.openInfoWindowHtml(point, msg);
				}						 		
			}
		);
		var request = getHTTPObject();	
		if(request){		
			request.onreadystatechange = function(){			
				parseMapResponse(request,map);		
			};		
			request.open("GET","data/map.php",false);	
			request.send(null);	
		}
	}
}
function parseMapResponse(request,map){	
	if(request.readyState == 4){
		if(request.status == 200 || request.status == 304){			
			var data=parseIEBug(request);
			var root=data.getElementsByTagName('Map')[0];
			if(root==null){return;}
			__items=root.getElementsByTagName("Grid");	
			if(__items==null){return;}
			for(var i=0;i<__items.length;i++){		
				if(__items[i].nodeType == 1){
					var xmluuid=__items[i].getElementsByTagName("Uuid")[0].firstChild.nodeValue;
					var xmlregionname=__items[i].getElementsByTagName("RegionName")[0].firstChild.nodeValue;
					var xmllocX=__items[i].getElementsByTagName("LocX")[0].firstChild.nodeValue;
					var xmllocY=__items[i].getElementsByTagName("LocY")[0].firstChild.nodeValue;
					xmllocX=xmllocX - xstart + 10;
					xmllocY=xmllocY - ystart + 10;
					boundaries=new GLatLngBounds(new GLatLng(xmllocY,xmllocX), new GLatLng(xmllocY+1,xmllocX+1)); 		
					var rx=new RegExp("(-)", "g");
					xmluuid = xmluuid.replace(rx,"");		
					layer=new GGroundOverlay('data/regions/' + xmluuid + '.jpg', boundaries);
					map.addOverlay(layer);
				}
			}	
		}
	}
}
function getRegionInfos(x,y){
	if(__items==null){return;}
	var response="";
	for(var i=0;i<__items.length;i++){		
		if(__items[i].nodeType == 1){
			var xmllocX=__items[i].getElementsByTagName("LocX")[0].firstChild.nodeValue;
			var xmllocY=__items[i].getElementsByTagName("LocY")[0].firstChild.nodeValue;			
			if(xmllocX==x&&xmllocY==y){
				var xmluuid=__items[i].getElementsByTagName("Uuid")[0].firstChild.nodeValue;				
				var xmlregionname=__items[i].getElementsByTagName("RegionName")[0].firstChild.nodeValue;	
				var rx=new RegExp("(-)", "g");
				xmluuid = xmluuid.replace(rx,"");
				response="<table>";
				response+="<tr><td><span id='name'><b>" + xmlregionname + "</b></span><br /></td></tr>";
				response+="<tr><td><span id='loc'>" + xmllocX + ", " + xmllocY + "</span></td></tr>";
   				response+="<tr><td><a class=\"add\" href=\"secondlife://"+hgdomain+":"+hgport+":"+xmlregionname+"/"+xjump+"/"+yjump+"/\">Hypergrid teleport</a><br /></td></tr>";
				xmlregionname = xmlregionname.replace(" ","+");
				response+="<tr><td><a class=\"add\" href=\"secondlife://http|!!"+hgdomain+"|"+hgport+"+"+xmlregionname+"\">Viewer 3 HG URL</a><br /></td></tr>";
			}
		}	
	}	
	return response;
}
function isOutOfBounds(x,y){
if(x<xstart-30||x>xstart+30){return true;}
if(y<ystart-30||y>ystart+30){return true;}
return false;
}
function parseIEBug(request){	
if (document.implementation && document.implementation.createDocument){
xmlDoc = request.responseXML;
}else if (window.ActiveXObject){
var testandoAppend = document.createElement('xml');
testandoAppend.setAttribute('innerHTML',request.responseText);
testandoAppend.setAttribute('id','_formjAjaxRetornoXML');
document.body.appendChild(testandoAppend);
document.getElementById('_formjAjaxRetornoXML').innerHTML = request.responseText;
xmlDoc = document.getElementById('_formjAjaxRetornoXML');
}
return xmlDoc;
}
function getHTTPObject(){
var xhr = false;
if(window.XMLHttpRequest){
var xhr = new XMLHttpRequest();
}else if(window.ActiveXObject) {
try{
var xhr = new ActiveXObject("Msxml2.XMLHTTP");
}catch(e){
try{
var xhr = new ActiveXObject("Microsoft.XMLHTTP");
}catch(e){xhr=false;}
}
}
return xhr;
}
